### PR TITLE
[FLINK-22464][tests] Fix OperatorCoordinator test which is stalling with AdaptiveScheduler

### PR DESCRIPTION
**Pull request by rmetzger**
This PR fixes a test instability. See comment in the PR for what's going on.

### CODE REVIEW

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java [hunk 1]
```diff
@@ -219,6 +223,30 @@ public Long map(Long value) throws Exception {
...
+                                        // AdaptiveScheduler starts the CheckpointCoordinator right
+                                        // before deploying tasks (when slots are available
+                                        // already), hence tasks will start running almost
+                                        // immediately, and the checkpoint gets triggered too late
+                                        // (it won't be able to complete before the artificial
+                                        // failure from this test)
+                                        // When we detect AdaptiveScheduler, we artificially slow
+                                        // down
+                                        if (adaptiveSchedulerEnabled) {
+                                            Thread.sleep(1);
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java:248]: surely there is a more sane way to do this, like waiting for notifyCheckpointComplete to arrive before throwing the exception.

Comment (zentol) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java:248]: (I suppose the _source_ should wait until at least 1 checkpoint was taken before emitting the last value?)

Comment (rmetzger) [reply flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java:248]: Thanks for taking a look at the change! After an insane amount of time trying to figure out this problem, I was so excited that I forgot the golden rule for tests #1: don't use `Thread.sleep` 🤕 . I'll look into alternatives.

**Comment by rmetzger** [comment]
I pushed an additional commit. You'll see that I was not able to get rid of the sleep :(. I just made the sleeping much more specific.
Since its a FLIP-27 source, we can't easily track the checkpoint progress there.
I don't think I've really addressed your concern, but I'm a bit lost what to do. If you have a better idea, let me know and I can try to implement it.

**Comment by zentol** [comment]
> Since its a FLIP-27 source, we can't easily track the checkpoint progress there.

The `SourceReader` gets notified of completed checkpoints. Could you not modify the `TestingNumberSequenceSource` to return a custom reader that completes some shared (org.apache.flink.testutils.junit.SharedObjects) CompletableFuture within notifyCheckpointComplete?

**Comment by rmetzger** [comment]
Thanks a lot for your input. With this comment, I was able to get rid of the sleep. It's a bit different to what you proposed, but it works very well (at least locally ;) ).

File: flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java [hunk 2]
```diff
@@ -290,13 +311,16 @@ private void fullFillPendingRequests() {
     }
 
     private static class TestingNumberSequenceSource extends NumberSequenceSource {
+
         private static final long serialVersionUID = 1L;
 
         private final int numSplits;
+        private final long waitForCheckpointAfterMessages;
```
Comment (zentol) [inline flink-tests/src/test/java/org/apache/flink/runtime/operators/coordination/OperatorEventSendingCheckpointITCase.java:318]: nit: rename to `numAllowedMessageBeforeCheckpoint`

**Review by zentol — APPROVED** [review]
Nice, this looks a lot better.

**Comment by rmetzger** [comment]
Thanks a lot for your review! Merging ...

**Pull request closed without merging**